Fix SSR and prevent hydrate issues

ValentinH opened the pull request:
I have started to use this package today and I need to support SSR. However, after successfully making it worked on client, I noticed that it wasn't working properly on server-side (only tried with render prop).

Indeed, even by setting the `guessedBreakpoint` property, the `currentBreakpoint` in the render prop was always null on server.

My first step was to set the default `currentBreakpoint` state to  `this.calculateCurrentBreakpoint(this.props.guessedBreakpoint) || null` which was working properly on the server.

However, I found a second issue: even if the rendered breakpoint was `mobile` on the server (value of `guessedBreakpoint`), the first rendered breakpoint in the browser was `null`. This is why I have added the `this.calculateCurrentBreakpoint(global.window.innerWidth)` to the default state value when on the client. 

To make it worked, I had to replace all `this.state.breakpoints` by `this.props.breakpoints` because state wasn't already set when calling `calculateCurrentBreakpoint()`. As far as I understand, there wasn't really a reason to read the breakpoints from the state instead of the props.

I hope this makes sense 🙂

ValentinH commented:
Another approach would be to change  `guessedBreakpoint` and `defaultBreakpoint` to be string (that would need to be set to a breakpoint key).

Then, we could set the default state to:
```js
state = {
  breakpoints: this.props.breakpoints || {},
  screenWidth: global.window ? global.window.innerWidth : this.props.breakpoints[this.props.defaultBreakpoint],
  currentBreakpoint: global.window
    ? this.calculateCurrentBreakpoint(global.window.innerWidth)
    : this.props.guessedBreakpoint
      ? this.props.guessedBreakpoint
      : null,
}
```

Both approaches make the error handling in `componentDidMount()` useless since it would break directly. To fix that, we could set the default state in the constructor and move error logic before.

ehellman commented:
Hey @ValentinH ! Thanks for the PR!

Recently got an issue suggesting that we add tests (not in this PR, need to set up the suite), I'm also debating adding what the React team has named ["fixtures"](https://github.com/facebook/react/tree/master/fixtures) which is basically mini-projects used to debug. Might be useful for extending/validating functionality.

Would this have been useful to you when debugging this and creating this PR?

Starting a review now.

ValentinH commented:
What was a bit hard for me was not having React as a dev dependency. Indeed, to implement this PR, I used npm link of this lib in my project and it wasn't working without React. I also had issues with the npm lock file that got 90% overwritten due to differences in the way versions were defined. I'm using latest npm, don't know if that was the reason.

ehellman commented:
@ValentinH Hmmm, I have no idea about that lockfile issue, but it sounds worth investigating. As for having React as a dev dependency, we can add that no problem. Doesn't affect anything since we're transpiling and not bundling. Feel free to update your PR with that change if you want!

But yes, when developing locally I also use the `npm link` method. Mostly because I want to cover as many cases as possible (big sign that we need tests :D)

ehellman commented on modules/ReactBreakpoints.js:
```diff
@@ -49,8 +49,14 @@ class ReactBreakpoints extends React.Component {
+    screenWidth: global.window
+      ? global.window.innerWidth
+      : this.props.defaultBreakpoint,
+    currentBreakpoint: global.window
```
Please add a small comment above `currentBreakpoint: global.window ? ...` to describe what's going on here.

My goal with these comments is that even an inexperienced developer can read the code and figure out basically what is happening and why :)

ehellman commented:
Everything looks and works great. I'd just like you to add a comment as described - great work and thanks for the contribution!

ValentinH commented:
Sure, I think it would be even easier to read if I were doing all of this check in the constructor, what do you think?

Was there SSR working before for you with this config?

ValentinH commented:
@ehellman I've added a commit to show what I meant with the constructor. I think it's simpler to reason about. I've also moved error handling in it which makes it works also on server. 🙂 

Moreover, it also handles better the `defaultBreakpoint` now.

However, I still think that `guessedBreakpoint` and `defaultBreakpoint` types are misleading. Indeed, since `currentBreakpoint` is a string (the key of the current breakpoint), I would expect those 2 prop to be the same.

You would define your `ReactBreakpoints` like this:
```jsx
<ReactBreakpoints guessedBreakpoint="mobile" breakpoints={breakpoints}>
	<whatever />
</ReactBreakpoints>
```

It would also simplify the initialisation of the state because we wouldn't have to call `this.calculateCurrentBreakpoint()` for `guessedBreakpoint` and `defaultBreakpoint`.

ValentinH commented:
@ehellman any update on this? 🙂

ehellman commented:
@ValentinH Looking at this tomorrow (tuesday)! I've been away at a conference with the worst WiFi in the world! :(
